internal/olm/operator/internal/registry_pod.go

```diff
@@ -0,0 +1,278 @@
+			return nil, podLogs, fmt.Errorf("error verifying pod creation: %v: and fetching logs: %v", err, logErr)
+		}
+		return nil, podLogs, fmt.Errorf("error verifying pod creation: %v", err)
+	}
+
+	return pod, "", nil
+}
+
+// create returns a pod built from an index image or returns if an existing pod with the same name is found
+func (rp *RegistryPod) create(ctx context.Context) (*corev1.Pod, error) {
+	// call init() to set defaults and validate the RegistryPod struct
+	err := rp.init()
+	if err != nil {
+		return nil, fmt.Errorf("error in initializing and validating registry pod: %v", err)
+	}
+
+	// call build() to make the pod definition
+	pod, err := rp.build()
+	if err != nil {
+		return nil, fmt.Errorf("error in building registry pod: %v", err)
+	}
+
+	// Check if Pod already exists
+	tmpPod, err := rp.Kubeclient.CoreV1().Pods(pod.Namespace).Get(ctx, pod.Name, metav1.GetOptions{})
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			// Create pod in kubernetes cluster using the above pod definition and clientset
+			pod, err = rp.Kubeclient.CoreV1().Pods(pod.Namespace).Create(ctx, pod, metav1.CreateOptions{})
+			if err != nil {
+				return nil, fmt.Errorf("error creating registry pod: %v", err)
+			}
+		} else {
+			return nil, fmt.Errorf("error getting existing registry pod: %v", err)
+		}
+	} else {
+		return tmpPod, nil
+	}
+
+	return pod, nil
+}
+
+func (rp *RegistryPod) verifyPodRunning(ctx context.Context, pod *corev1.Pod) error {
+	// Upon creation of new pod, poll and verify that pod status is running
+	podCheck := wait.ConditionFunc(func() (done bool, err error) {
+		p, err := rp.Kubeclient.CoreV1().Pods(pod.Namespace).Get(ctx, pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, fmt.Errorf("error getting pod %s: %w", pod.Name, err)
+		}
+		return p.Status.Phase == corev1.PodRunning, nil
+	})
+
+	err := wait.PollImmediateUntil(time.Duration(200*time.Millisecond), podCheck, ctx.Done())
+	if err != nil {
+		return fmt.Errorf("error waiting for registry pod %s to run: %v", pod.Name, err)
+	}
+
+	return err
+}
+
+func (rp *RegistryPod) init() error {
+	rp.setDefaults()
+	err := rp.validate()
+	if err != nil {
+		return fmt.Errorf("error in validating registry pod: %v", err)
+	}
+	return nil
+}
+
+func (rp *RegistryPod) setDefaults() {
+	if rp.GRPCPort == 0 {
+		rp.GRPCPort = defaultGRPCPort
+	}
+
+	if len(rp.IndexImage) == 0 {
+		rp.IndexImage = defaultIndexImage
+	}
+
+	if len(rp.BundleAddMode) == 0 {
+		if rp.IndexImage == defaultIndexImage {
+			rp.BundleAddMode = "semver"
+		} else {
+			rp.BundleAddMode = "replaces"
+		}
+	}
+}
+
+func (rp *RegistryPod) validate() error {
+	if len(rp.BundleImage) == 0 {
+		return errors.New("bundle image is required")
+	}
+	if len(rp.DBPath) == 0 {
+		return errors.New("registry database path is required")
+	}
+
+	if len(rp.Namespace) == 0 {
+		return errors.New("pod namespace is required")
+	}
+
+	if len(rp.BundleAddMode) == 0 {
+		panic("bundle add mode is not set")
+	}
```

Comment:
```suggestion
```

We should NOT do this check and the validation IMO should be called before set the defaults.
First, we validate data inputted by the user and then, if all is ok we work with.